test/metabase/test/data/drill.clj

```diff
@@ -0,0 +1,188 @@
+   :type/Decimal    "DECIMAL"
+   :type/Float      "DOUBLE"
+   :type/Integer    "INTEGER"
+   :type/Text       "VARCHAR"})
+
+(defn- dash-to-underscore [s]
+  (s/replace s #"-" "_"))
+
+(defn- qualified-name-components
+  ([db-name]                       (map dash-to-underscore [db-name]))
+  ([db-name table-name]            (map dash-to-underscore [db-name table-name]))
+  ([db-name table-name field-name] (map dash-to-underscore [table-name field-name])))
+
+(defn database->connection-details [context {:keys [database-name]}]
+  (merge {:drill-connect "drillbit=localhost;schema=dfs.tmp"}))
+
+(defn quote-name [nm]
+  (str \` nm \`))
+
+(defn make-sure-file-exists! [path]
+  (let [file-path (clojure.java.io/file path)]
+    (when-not (.exists file-path)
+      (.createNewFile file-path))))
+
+(defn create-table-sql [{:keys [database-name], :as dbdef} {:keys [table-name field-definitions]}]
+  (let [qualified-table-name (i/db-qualified-table-name database-name table-name)
+        table-path (str "/tmp/" qualified-table-name "_table.csv")]
+    ;; we can't create the view unless the file exists, so create an empty one if necessary
+    (make-sure-file-exists! table-path)
+    (format "CREATE OR REPLACE VIEW %s AS SELECT %s FROM dfs.`%s`"
+            (str "dfs.tmp.`" qualified-table-name "`")
+            (str (->> field-definitions
+                      (map (fn [{:keys [field-name base-type]}]
+                             (if (= base-type :type/DateTime)
+                               (format "CAST(to_timestamp(`%s`, 'YYYY-MM-dd''T''HH:mm:ss.SSSZ') AS TIMESTAMP) AS `%s`"
+                                       field-name field-name)
+                               (format "CAST(`%s` AS %s) AS `%s`"
+                                       field-name
+                                       (field-base-type->sql-type base-type)
+                                       field-name))))
+                      (interpose ", ")
+                      (apply str))
+                 ", CAST(`id` as INTEGER) AS `id` ")
+            table-path)))
+
+(defn drop-table-if-exists-sql [{:keys [database-name]} {:keys [table-name]}]
+  (format "DROP VIEW IF EXISTS %s" (str "dfs.tmp.`" database-name "_" table-name "`")))
+
+;; workaround for DRILL-5136 (can't use prepared statements for DDL)
+;; instead of prepared statements, we use plain statements.
+;; since execute in the jdbc driver uses prepared statements by default,
+;; we use .createStatement and .execute explicitly.
+(defn execute-without-prepared-statement! [conn driver context dbdef sql]
+  (let [sql (some-> sql s/trim)]
+    (when (and (seq sql)
+               ;; make sure SQL isn't just semicolons
+               (not (s/blank? (s/replace sql #";" ""))))
+      ;; Remove excess semicolons, otherwise snippy DBs like Oracle will barf
+      (let [sql (s/replace sql #";+" ";")]
+        (try
+          (with-open [sql-statement (.createStatement conn)]
+            (.execute sql-statement sql))
```

Comment:
It looks like this part is the only part that's different from the `default-execute-sql!` implementation. It would be a lot nicer just to tweak the original function for flexibility rather than copying the whole thing over to change one thing.

One thing we could do is let it take an optional `execute` function the way `sequentially-execute-sql!` does:

```clojure
(defn- jdbc-execute! [db-spec sql]
  (jdbc/execute! db-spec [sql] {:transaction? false, :multi? true}))

(defn default-execute-sql! [driver context dbdef sql & {:keys [execute]
                                                        :or {execute jdbc-execute!}}]
  (let [sql (some-> sql s/trim)]
    (when (and (seq sql)
               ;; make sure SQL isn't just semicolons
               (not (s/blank? (s/replace sql #";" ""))))
      ;; Remove excess semicolons, otherwise snippy DBs like Oracle will barf
      (let [sql (s/replace sql #";+" ";")]
        (try
          (execute db-spec sql)
          (catch SQLException e
            (println "Error executing SQL:" sql)
            (printf "Caught SQLException:\n%s\n"
                    (with-out-str (jdbc/print-sql-exception-chain e)))
            (throw e))
          (catch Throwable e
            (println "Error executing SQL:" sql)
            (printf "Caught Exception: %s %s\n%s\n" (class e) (.getMessage e)
                    (with-out-str (.printStackTrace e)))
            (throw e)))))))
```

The this whole implementation here could look like

```clojure
(def ^:dynamic ^:private *connection* nil)

;; workaround for DRILL-5136 (can't use prepared statements for DDL)
;; instead of prepared statements, we use plain statements.
;; since execute in the jdbc driver uses prepared statements by default,
;; we use .createStatement and .execute explicitly.
(def ^:private execute-without-prepared-statement! [driver context dbdef sql]
  (generic/default-execute-sql! driver context dbdef sql
                                :execute (fn [_ sql]
                                           (with-open [sql-statement (.createStatement *connection*)]
                                             (.execute sql-statement sql)))))

(defn sequentially-execute-sql-without-prepared-statements! [driver context dbdef sql]
  (with-open [conn (jdbc/get-connection (generic/database->spec driver context dbdef))]
    (binding [*connection* conn]
      (generic/sequentially-execute-sql! driver context dbdef sql
                                         :execute (partial execute-without-prepared-statement! conn)))))
```

That would save us 30 lines of copy-pasta or so

Reply:
Alternatively even just abstracting out the error handling and logging so it can be reused would be a huge win 

Keeping code duplication to a minimum here is a good thing because if we can make the normal driver implementations more flexible/easier to assemble together it will make maintainence and writing future drivers easier